File: c/quickSort.c
//selection sort
#include <stdio.h>

void swap(int *xp, int *yp){
    int temp = *xp;
    *xp = *yp;
    *yp = temp;
}

int partition (int arr[], int low, int high){
    int pivot = arr[high]; 
    int i = (low - 1); 
    for (int j = low; j <= high - 1; j++){
        if (arr[j] < pivot){
            i++; 
            swap(&arr[i], &arr[j]);
        }
    }
    swap(&arr[i + 1], &arr[high]);
    return (i + 1);
}


void quickSort(int arr[], int low, int high){
    if (low < high){  
        int pi = partition(arr, low, high);

        quickSort(arr, low, pi - 1);
        quickSort(arr, pi + 1, high);
    }
}

int main(){
    int arr[]={98,45,78,32,99,12,30,26,10};
    int n = sizeof(arr) / sizeof(arr[0]);
    printf("Unsorted array:\n");
    for(int i=0;i<n;i++)
        printf("%d ",arr[i]);

    quickSort(arr,0,n);    
    printf("\nSorted array:\n");
    for(int i=0;i<n;i++)
        printf("%d ",arr[i]);    
    return 0;
}

File: go/quickSort.go
package main

import "fmt"

func partition(A []int, low, high int) int {
	pivot := A[high]
	i := low - 1

	for j := low; j <= high-1; j++ {
		if A[j] < pivot {
			i++
			A[i], A[j] = A[j], A[i]
		}
	}

	A[i+1], A[high] = A[high], A[i+1]

	return i + 1
}

func quickSort(A []int, low, high int) {
	if low < high {
		p := partition(A, low, high)

		quickSort(A, low, p-1)
		quickSort(A, p+1, high)
	}
}

func main() {
	arr := []int{
		147, 485, 735, 606, 280,
		243, 810, 282, 940, 314,
		237, 576, 498, 876, 555,
		656, 573, 513, 937, 289,
		2, 678, 653, 833, 333,
	}

	fmt.Printf("Unsorted: %v\n", arr)

	quickSort(arr, 0, len(arr)-1)

	fmt.Printf("Sorted:   %v\n", arr)
}
